java/ql/lib/semmle/code/java/Statement.qll

```diff
@@ -483,7 +527,58 @@ class ConstCase extends SwitchCase {
   override string getAPrimaryQlClass() { result = "ConstCase" }
 }
 
-/** A `default` case of a `switch` statement */
+/**
+ * A binding or record pattern.
+ *
+ * Note binding patterns are represented as `LocalVariableDeclExpr`s.
+ */
+class PatternExpr extends Expr {
+  PatternExpr() {
+    (
+      this.getParent() instanceof SwitchCase or
+      this.getParent() instanceof InstanceOfExpr or
+      this.getParent() instanceof PatternExpr
```

Comment:
Couldn't we change this to `RecordPatternExpr` without changing semantics, but thereby getting rid of recursion?
```suggestion
      this.getParent() instanceof RecordPatternExpr
```